worker/src/main/scala/org/apache/celeborn/service/deploy/worker/storage/DeviceMonitor.scala

```diff
@@ -225,26 +235,43 @@ class LocalDeviceMonitor(
+              if (nonCriticalErrorSum > device.notifyErrorThreshold) {
+                logger.error(s"Device ${device.deviceInfo.name} has accumulated $nonCriticalErrorSum non-critical " +
+                  s"error within the past ${Utils.msDurationToString(device.notifyErrorExpireTimeout)} , its sum has " +
+                  s"exceed the threshold (${device.notifyErrorThreshold}), device monitor will notify error to " +
+                  s"observed device.")
+                val mountPoints = device.diskInfos.values().asScala.map(_.mountPoint).toList
+                device.notifyObserversOnError(mountPoints, DiskStatus.CRITICAL_ERROR)
               } else {
-                device.diskInfos.values().asScala.foreach { case diskInfo =>
-                  if (checkDiskUsage && DeviceMonitor.highDiskUsage(conf, diskInfo)) {
-                    logger.error(s"${diskInfo.mountPoint} high_disk_usage error, notify observers")
-                    device.notifyObserversOnHighDiskUsage(diskInfo.mountPoint)
-                  } else if (checkReadWrite &&
-                    DeviceMonitor.readWriteError(conf, diskInfo.dirs.head)) {
-                    logger.error(s"${diskInfo.mountPoint} read-write error, notify observers")
-                    // We think that if one dir in device has read-write problem, if possible all
-                    // dirs in this device have the problem
-                    device.notifyObserversOnNonCriticalError(
-                      List(diskInfo.mountPoint),
-                      DiskStatus.READ_OR_WRITE_FAILURE)
-                  } else {
-                    device.notifyObserversOnHealthy(diskInfo.mountPoint)
+                if (checkIoHang && device.ioHang()) {
+                  logger.error(s"Encounter device io hang error!" +
+                    s"${device.deviceInfo.name}, notify observers")
+                  device.notifyObserversOnNonCriticalError(mountPoints, DiskStatus.IO_HANG)
+                } else {
+                  device.diskInfos.values().asScala.foreach { case diskInfo =>
+                    if (checkDiskUsage && DeviceMonitor.highDiskUsage(conf, diskInfo)) {
+                      logger.error(
+                        s"${diskInfo.mountPoint} high_disk_usage error, notify observers")
+                      device.notifyObserversOnHighDiskUsage(diskInfo.mountPoint)
+                    } else if (checkReadWrite &&
+                      DeviceMonitor.readWriteError(conf, diskInfo.dirs.head)) {
+                      logger.error(s"${diskInfo.mountPoint} read-write error, notify observers")
+                      // We think that if one dir in device has read-write problem, if possible all
+                      // dirs in this device have the problem
+                      device.notifyObserversOnNonCriticalError(
+                        List(diskInfo.mountPoint),
+                        DiskStatus.READ_OR_WRITE_FAILURE)
+                    } else if (nonCriticalErrorSum == 0) {
```

Comment:
Consider this situation: at time 1 disk1 checkreadwirte fail, then the status is set to READ_OR_WRITE_FAILURE and add the timestamp into  nonCriticalErrors; at time 2 disk1 checkreadwrite passed, but since nonCriticalErrorSum is not 0(time not expired), it will not become healthy.

Reply:
Yes, it will also be a problem. How about `nonCriticalErrorSum <= device.notifyErrorThreshold/2`? 
There will be 3 scenarios:
1. `nonCriticalErrorSum ∈ [0, device.notifyErrorThreshold/2]` Disk status will be healthy or nonCriticalError like READ_OR_WRITE_FAILURE, and it could also recover to healthy once there isn't new nonCriticalError caught.
2. `nonCriticalErrorSum ∈ (device.notifyErrorThreshold/2, device.notifyErrorThreshold]`, Disk status will be nonCriticalError, and it will affect the allocating new slots and creating new fileWriter on this disk because its status can't become back to healthy, but won't notify and close all related DeviceObserver.
3. `nonCriticalErrorSum ∈ (device.notifyErrorThreshold, +♾)`, status will be CRITICAL_ERROR, then stop and close all related DeviceObserver.
BTW can we make this ratio 1/2 as a config, which should ∈[0,1] and set its default value as 0.5?
cc: @waitinfuture @AngersZhuuuu

Reply:
I think this design is fine, thanks for the detailed description. BTW, it's interesting how you typed ♾